### PR TITLE
types: add deepcopy for update operation

### PR DESCRIPTION
**Pull request by mccxj**
for #5064 #5084 
Add 'deepcopy' maybe not suitable, PTAL.

### CODE REVIEW

**Comment by jackysp** [comment]
Thank you for this quick fix. Please add some test cases, for example JSON update case which is mentioned in #5064 .

**Comment by jackysp** [comment]
LGTM. Please add some test cases.

**Comment by jackysp** [comment]
PTAL @hicqu @winoros @wentaoxu

**Comment by shenli** [comment]
LGTM

File: types/datum.go [hunk 1]
```diff
@@ -64,6 +64,19 @@ type Datum struct {
...
 }
 
+// Copy deepcopy a Datum.
+func (d *Datum) Copy() *Datum {
+	ret := *d
+	if d.b != nil {
+		ret.b = make([]byte, len(d.b))
+		copy(ret.b, d.b)
+	}
+	if ret.k == KindMysqlJSON {
```
Comment (XuHuaiyu) [inline types/datum.go]: add a comment for this check.

File: types/datum.go [hunk 2]
```diff
@@ -64,6 +64,20 @@ type Datum struct {
 	x         interface{} // x hold all other types.
 }
 
+// Copy deepcopy a Datum.
```
Comment (zz-jason) [inline types/datum.go]: s/deepcopy/deep copies/

File: types/json/json.go [hunk 3]
```diff
@@ -67,6 +67,24 @@ type JSON struct {
 	Array    []JSON
 }
 
+// Copy deepcopy a JSON.
```
Comment (zz-jason) [inline types/json/json.go]: ditto

File: types/datum.go (hunk 2 above)
Comment (mccxj) [reply types/datum.go]: OK. PTAL.

**Pull request merged**